File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/Arithmetics/Interval.scala
package OSOL.Extremum.Core.Scala.Arithmetics

import Interval.Exceptions._

/** Represents interval numbers in the form `[a; b]`
  *
  * @param lowerBound min value `a` of the interval
  * @param upperBound max value `b` of the interval
  *
  */
class Interval private (val lowerBound: Double, val upperBound: Double) {

  /** Gets middle point if the interval */
  final lazy val middlePoint: Double = 0.5 * (lowerBound + upperBound)
  /** Gets width of the interval */
  final lazy val width: Double = upperBound - lowerBound
  /** Gets radius of the interval */
  final lazy val radius: Double = width / 2.0

  /** Converts Interval to String
    *
    * @return string representation of Interval
    */
  final override def toString: String = s"[$lowerBound; $upperBound]"

  /** Deremines whether objects are equal or not
    *
    * @param that second object
    * @return equal or not
    */
  final def equalsTo(that: Interval): Boolean =
    this.lowerBound == that.lowerBound && this.upperBound == that.upperBound
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#equalsTo equalsTo]] */
  final def ==(that: Interval): Boolean = this.equalsTo(that)

  /** Deremines whether objects are approximately equal
    *
    * @param that second object
    * @param maxError maximum allowed difference
    * @return equal or not
    */
  final def approximatelyEqualsTo(that: Interval, maxError:Double): Boolean = {
    def getDifference(a: Double, b: Double): Double = {
      val delta = math.abs(a - b)
      if (delta.isNaN) {
        if (a == b) 0.0
        else 1.0
      }
      else delta
    }
    val errorLeft = getDifference(this.lowerBound, that.lowerBound)
    val errorRight = getDifference(this.upperBound, that.upperBound)
    errorLeft + errorRight <= maxError
  }
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#approximatelyEqualsTo approximatelyEqualsTo]] */
  final def ~(that: Interval, maxError: Double = 1e-5): Boolean = this.approximatelyEqualsTo(that, maxError)

  /** Negates current Interval `[a; b]`
    *
    * @return `[-b; -a]`
    */
  final def neg(): Interval =
    Interval(-upperBound, -lowerBound)
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#neg neg]] */
  final def unary_-(): Interval = this.neg()

  /** Calculates sum of two intervals
    *
    * @param that second term
    * @return sum of `this + that`
    */
  final def add(that: Interval): Interval =
    Interval(this.lowerBound + that.lowerBound, this.upperBound + that.upperBound)
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#add add]] */
  final def +(that: Interval): Interval = this.add(that)

  /** Calculates difference of two intervals
    *
    * @param that second term
    * @return sum of `this - that`
    */
  final def sub(that: Interval): Interval =
    Interval(this.lowerBound - that.upperBound, this.upperBound - that.lowerBound)
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#sub sub]] */
  final def -(that: Interval): Interval = this.sub(that)

  /** Calculates product of two intervals
    *
    * @param that second term
    * @return sum of `this * that`
    */
  final def mul(that: Interval): Interval = {
    val products = Seq(
      this.lowerBound * that.lowerBound,
      this.lowerBound * that.upperBound,
      this.upperBound * that.lowerBound,
      this.upperBound * that.upperBound)

    Interval(products.min, products.max)
  }
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#mul mul]] */
  final def *(that: Interval): Interval = this.mul(that)

  /** Calculates quotient of two intervals
    *
    * @param that second term
    * @return sum of `this / that`
    */
  final def div(that: Interval): Interval = {
    that match {
      case _ if that.lowerBound > 0.0 || that.upperBound < 0.0 => this.mul(Interval(1.0 / that.upperBound, 1.0 / that.lowerBound))
      case _ if that.lowerBound == 0.0 => this.mul(Interval(1.0 / that.upperBound, Double.PositiveInfinity))
      case _ if that.upperBound == 0.0 => this.mul(Interval(Double.NegativeInfinity, 1 / that.lowerBound))
      case _ => Interval(Double.NegativeInfinity, Double.PositiveInfinity)
    }
  }
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#div div]] */
  final def /(that: Interval): Interval = this.div(that)

  /** Calculates power of two intervals
    *
    * @param that second term
    * @return sum of `this \^ that`
    */
  final def pow(that: Interval): Interval = {
    if (that.width != 0) throw new UnknownOperationException("[a; b] ^ [c; d] for (d - c) > 0")
    else {
      val powerIndex: Int = that.middlePoint.toInt
      val values = Seq(math.pow(lowerBound, powerIndex), math.pow(upperBound, powerIndex)).sorted
      (that, values) match {
        case (_, a :: b :: _) if powerIndex % 2 == 0 => if (lowerBound * upperBound <= 0.0) Interval(0.0, b) else Interval(a, b)
        case (_, a :: b :: _) => Interval(a, b)
      }
    }
  }
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#pow pow]] */
  final def ^(that: Interval): Interval = this.pow(that)
  /** Same as [[OSOL.Extremum.Core.Scala.Arithmetics#pow pow]] */
  final def **(that: Interval): Interval = this.pow(that)

  /** Calculates `abs(this)`
    *
    * @return absolute value
    */
  final def abs(): Interval = {
    val values = Seq(math.abs(lowerBound), math.abs(upperBound)).sorted
    values match {
      case _ :: b :: _ if lowerBound * upperBound <= 0.0 => Interval(0.0, b)
      case a :: b :: _ => Interval(a, b)
    }
  }

  /** Calculates `sin(this)`
    *
    * @return sine value
    */
  final def sin(): Interval = {
    if (width > 2.0 * math.Pi) Interval(-1.0, 1.0)
    else {
      val c = 0.5 * math.Pi
      val leftBound = math.ceil(lowerBound / c).toInt
      val rightBound = math.floor(upperBound / c).toInt
      val points = lowerBound +: upperBound +: (leftBound to rightBound).map(c * _)
      val mappedPoints = points.map(math.sin).sorted
      Interval(mappedPoints.head, mappedPoints.last)
    }
  }

  /** Calculates `cos(this)`
    *
    * @return cosine value
    */
  final def cos(): Interval = {
    if (width > 2.0 * math.Pi) Interval(-1.0, 1.0)
    else {
      val c = 0.5 * math.Pi
      val leftBound = math.ceil(lowerBound / c).toInt
      val rightBound = math.floor(upperBound / c).toInt
      val points = lowerBound +: upperBound +: (leftBound to rightBound).map(c * _)
      val mappedPoints = points.map(math.cos).sorted
      Interval(mappedPoints.head, mappedPoints.last)
    }
  }

  /** Calculates `exp(this)`
    *
    * @return exponent value
    */
  final def exp(): Interval = Interval(math.exp(lowerBound), math.exp(upperBound))

  /** Calculates `sqrt(this)`
    *
    * @return square root value
    */
  final def sqrt(): Interval = {
    if (upperBound < 0.0) throw new BadAreaOperationException(opName = "sqrt", this)
    else {
      if (lowerBound < 0.0) Interval(0.0, math.sqrt(upperBound))
      else Interval(math.sqrt(lowerBound), math.sqrt(upperBound))
    }
  }

  /** Calculates `ln(this)`
    *
    * @return natural logarithm value
    */
  final def ln(): Interval = {
    if (upperBound < 0.0) throw new BadAreaOperationException(opName = "ln", this)
    else {
      if (lowerBound < 0.0) Interval(Double.NegativeInfinity, math.log(upperBound))
      else Interval(math.log(lowerBound), math.log(upperBound))
    }
  }

  /** Moves current interval according to `delta` value
    *
    * @param delta movement value
    * @return moved interval
    */
  final def moveBy(delta: Double): Interval = this.add(Interval(delta))

  /** Forces interval to be located in `[min; max]` interval
    *
    * @param min minimum allowed value
    * @param max maximum allowed value
    * @return constrained interval
    */
  final def constrain(min: Double, max: Double): Interval = {
    def check(value: Double) =
      if (value < min) min
      else {
        if (value > max) max
        else value
      }
    Interval(check(lowerBound), check(upperBound))
  }

}

/** Companion object for [[OSOL.Extremum.Core.Scala.Arithmetics.Interval Interval]] */
object Interval {

  import Converters._

  /** Minimal width to consider interval */
  var minWidth: Double = 1e-5

  /** Constructor that analyses width
    *
    * @param lowerBound min value of interval
    * @param upperBound max value of interval
    * @return degenerate interval (if `upperBound` - `lowerBound` < `minWidth`) or <p>
    *         ordinary interval ['lowerBound', 'upperBound']
    */
  final def apply(lowerBound: Double, upperBound: Double): Interval = {
    if (upperBound - lowerBound < minWidth) Interval(0.5 * (lowerBound + upperBound))
    else new Interval(lowerBound, upperBound)
  }

  /** Create Interval number from point
    *
    * @param value midpoint
    * @return degenerate interval [`value`; `value`]
    **/
  final def apply(value: Double): Interval = value

  /** Implicit converters for [[OSOL.Extremum.Core.Scala.Arithmetics.Interval Interval]] */
  object Converters {

    /** Converts Double value to Interval number
      *
      * @param value value to convert
      * */
    implicit def fromDouble(value: Double): Interval = new Interval(value, value)

    /** Converts String value to Interval number
      *
      * @param value value to convert
      * */
    implicit def fromString(value: String): Interval = {
      val split = value.drop(1).dropRight(1).split(";")
      if (split.length == 2) Interval(split(0).toDouble, split(1).toDouble)
      else fromDouble(value.toDouble)
    }

    /** Converts pair of values (a, b) to Interval number
      *
      * @param pair value to convert
      * @return interval [`pair_1`; `pair_2`]
      * */
    implicit def fromPair(pair: (Double, Double)): Interval = Interval(pair._1, pair._2)

  }

  /** Exceptions that can be raised for [[OSOL.Extremum.Core.Scala.Arithmetics.Interval Interval]] */
  object Exceptions {

    /** Exception for unsupported operation
      *
      * @param opName operation name
      */
    class UnknownOperationException(opName: String) extends Exception

    /** Exception for bad operation argument
      *
      * @param opName operation name
      * @param interval bad interval
      */
    class BadAreaOperationException(opName: String, interval: Interval) extends Exception

  }

}

File: Cores/OSOL.Extremum.Core.DotNet/Arithmetics/Interval.cs
using System;
using System.Linq;
using System.Net.Sockets;
using System.Reflection.Metadata.Ecma335;

namespace OSOL.Extremum.Core.DotNet.Arithmetics
{
    public struct Interval
    {
        public static double MinWidth = 1e-5;
        
        public double LowerBound, UpperBound;
        public double MiddlePoint => 0.5 * (UpperBound + LowerBound);
        public double Width => UpperBound - LowerBound;
        public double Radius => 0.5 * Width;

        public Interval(double value)
        {
            this.LowerBound = this.UpperBound = value;
        }
        public Interval(double lowerBound, double upperBound)
        {
            double w = upperBound - lowerBound;
            if (w < MinWidth)
            {
                this.LowerBound = this.UpperBound = 0.5 *(lowerBound + upperBound);
            }
            else
            {
                this.LowerBound = lowerBound;
                this.UpperBound = upperBound;
            }
        }
        
        public static implicit operator Interval(double value) => new Interval(value);

        public override string ToString() => $"[{LowerBound}; {UpperBound}]";

        public bool EqualsTo(Interval that) => this.LowerBound == that.LowerBound && this.UpperBound == that.UpperBound;

        public static bool operator ==(Interval a, Interval b) => a.EqualsTo(b);
        public static bool operator !=(Interval a, Interval b) => !(a == b);

        public bool ApproximatelyEqualsTo(Interval that, double maxError = 1e-5)
        {
            Func<double, double, double> getDifference = (a, b) =>
            {
                double delta = Math.Abs(b - a);
                if (double.IsNaN(delta)) return a == b ? 0.0 : 1.0;
                else return delta;
            };
            double errorLeft = getDifference(this.LowerBound, that.LowerBound);
            double errorRight = getDifference(this.UpperBound, that.UpperBound);
            return errorLeft + errorRight <= maxError;
        }
        
        public Interval Neg() => new Interval(-UpperBound, -LowerBound);
        public static Interval operator -(Interval i) => i.Neg();
        
        public Interval Add(Interval that) => new Interval(this.LowerBound + that.LowerBound, this.UpperBound + that.UpperBound);
        public static Interval operator +(Interval a, Interval b) => a.Add(b);

        public Interval Subtract(Interval that) => this.Add(that.Neg());
        public static Interval operator -(Interval a, Interval b) => a.Subtract(b);

        public Interval Multiply(Interval that)
        {
            double[] values = new double[]
            {
                this.LowerBound * that.LowerBound,
                this.LowerBound * that.UpperBound,
                this.UpperBound * that.LowerBound,
                this.UpperBound * that.UpperBound
            };
            return new Interval(values.Min(), values.Max());
        }
        public static Interval operator *(Interval a, Interval b) => a.Multiply(b);

        public Interval Divide(Interval that)
        {
            if (that.LowerBound > 0.0 || that.UpperBound < 0.0)
                return this.Multiply(new Interval(1.0 / that.UpperBound, 1.0 / that.LowerBound));
            if (that.LowerBound == 0.0)
                return this.Multiply(new Interval(1.0 / that.UpperBound, double.PositiveInfinity));
            if (that.UpperBound == 0.0)
                return this.Multiply(new Interval(double.NegativeInfinity, 1 / that.LowerBound));
            return new Interval(double.NegativeInfinity, double.PositiveInfinity);
        }
        public static Interval operator /(Interval a, Interval b) => a.Divide(b);

        public Interval Power(Interval that)
        {
            if(that.Width > 0) throw new IntervalExceptions.UnknownOperationException("[a; b] ^ [c; d] for (d - c) > 0");
            else
            {
                int powerIndex = (int)that.MiddlePoint;
                double v1 = Math.Pow(this.LowerBound, powerIndex);
                double v2 = Math.Pow(this.UpperBound, powerIndex);
                if(powerIndex % 2 == 1)
                    return new Interval(v1, v2);
                else
                {
                    if(this.LowerBound * this.UpperBound < 0) return new Interval(0.0, Math.Max(v1, v2));
                    else return this.LowerBound >= 0 ? new Interval(v1, v2) : new Interval(v2, v1);
                }
            }
        }

        public Interval Abs()
        {
            double v1 = Math.Abs(this.LowerBound);
            double v2 = Math.Abs(this.UpperBound);
            if(this.LowerBound * this.UpperBound < 0) return new Interval(0.0, Math.Max(v1, v2));
            else return this.LowerBound >= 0 ? new Interval(v1, v2) : new Interval(v2, v1);
        }
        
        public Interval Sin()
        {
            if (this.Width > Math.PI * 2.0)
                return new Interval(-1, 1);
            double val, k, r_l = Math.Sin(this.LowerBound), r_u = Math.Sin(this.UpperBound);
            if (r_u < r_l)
            {
                k = r_l;
                r_l = r_u;
                r_u = k;
            }
            k = (int)(this.LowerBound / (Math.PI / 2.0));
            val = k * (Math.PI / 2.0);
            while (val < this.LowerBound)
            {
                ++k;
                val = k * (Math.PI / 2.0);
            }
            for (; val <= this.UpperBound; ++k, val = k * (Math.PI / 2.0))
            {
                if (r_l == -1 && r_u == 1)
                    break;
                else
                {
                    if (Math.Sin(val) == -1)
                    {
                        r_l = -1;
                        continue;
                    }
                    if (Math.Sin(val) == 1)
                    {
                        r_u = 1;
                        continue;
                    }
                }
            }
            return new Interval(r_l, r_u);
        }

        public Interval Cos()
        {
            if (this.Width > Math.PI * 2.0)
                return new Interval(-1, 1);
            double val, k, r_l = Math.Cos(this.LowerBound), r_u = Math.Cos(this.UpperBound);
            if (r_u < r_l)
            {
                k = r_l;
                r_l = r_u;
                r_u = k;
            }
            k = (int)(this.LowerBound / (Math.PI / 2.0));
            val = k * (Math.PI / 2.0);
            while (val < this.LowerBound)
            {
                ++k;
                val = k * (Math.PI / 2.0);
            }
            for (; val <= this.UpperBound; ++k, val = k * (Math.PI / 2.0))
            {
                if (r_l == -1 && r_u == 1)
                    break;
                else
                {
                    switch (Math.Cos(val))
                    {
                        case -1:
                            r_l = -1;
                            continue;
                        case 1:
                            r_u = 1;
                            continue;
                    }
                }
            }
            return new Interval(r_l, r_u);
        }

        public Interval Exp() => new Interval(Math.Exp(this.LowerBound), Math.Exp(this.UpperBound));

        public Interval Sqrt()
        {
            if (this.UpperBound < 0)
                throw new IntervalExceptions.BadAreaOperationException(opName: "Ln", interval: this);
            else
            {
                if (this.LowerBound < 0) return new Interval(0.0, Math.Sqrt(this.UpperBound));
                else return new Interval(Math.Sqrt(this.LowerBound), Math.Sqrt(this.UpperBound));
            }  
        }

        public Interval Ln()
        {
            if (this.UpperBound < 0)
                throw new IntervalExceptions.BadAreaOperationException(opName: "Ln", interval: this);
            else
            {
                if (this.LowerBound < 0) return new Interval(double.NegativeInfinity, Math.Log(this.UpperBound));
                else return new Interval(Math.Log(this.LowerBound), Math.Log(this.UpperBound));
            }  
        }
        
    }
}